src/transformation/visitors/jsx/jsx.ts

```diff
@@ -0,0 +1,252 @@
+import * as ts from "typescript";
+import { JsxEmit } from "typescript";
+import * as lua from "../../../LuaAST";
+import { FunctionVisitor, TransformationContext, Visitors } from "../../context";
+import { transformJsxAttributes } from "../literal";
+import { unsupportedJsxConfiguration } from "../../utils/diagnostics";
+import { XHTMLEntities } from "./xhtml";
+import { AnnotationKind, getFileAnnotations } from "../../utils/annotations";
+
+function checkValidJsxConfig(node: ts.Node, context: TransformationContext) {
+    if (context.options.jsx !== JsxEmit.React) {
+        context.diagnostics.push(unsupportedJsxConfiguration(node));
+    }
+}
+
+// used for jsxFactory and jsxFragmentFactory options
+function getExpressionFromOption(
+    node: ts.Node,
+    fileAnnotation: AnnotationKind,
+    option: string | undefined,
+    defaultValue: string
+): lua.Expression {
+    const annotation = getFileAnnotations(node.getSourceFile()).get(fileAnnotation);
+    const [first, second] = (annotation?.args[0] ?? option ?? defaultValue).split(".");
+    return second === undefined
+        ? lua.createIdentifier(first)
+        : lua.createTableIndexExpression(lua.createIdentifier(first), lua.createStringLiteral(second));
+}
+
+/* Implementations of jsx text processing modified from sucrase (https://github.com/alangpierce/sucrase). */
```

Comment:
Please include the full licence of sucrase here in a comment.

Reply:
sucrase uses the MIT licence, and I copied some portions of their implementation.
How would I properly attribute such code? Still not too familiar with the particulars of OSS licensing.

Reply:
Full text of license was added.

Reply:
The correct way would be to add the license to our binary as well e.g. to `tstl -h`. But adding the full text or a permalink in code is good enough I think. That's what we've done in other places where third party code was used. As long as no one complains this should be fine.

thanks for updating.